rskj-core/src/main/java/co/rsk/pcc/NativeContract.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2019 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.pcc;
+
+import co.rsk.config.RskSystemProperties;
+import co.rsk.core.RskAddress;
+import co.rsk.panic.PanicProcessor;
+import org.bouncycastle.util.encoders.Hex;
+import org.ethereum.core.Block;
+import org.ethereum.core.Repository;
+import org.ethereum.core.Transaction;
+import org.ethereum.db.BlockStore;
+import org.ethereum.db.ReceiptStore;
+import org.ethereum.vm.LogInfo;
+import org.ethereum.vm.PrecompiledContracts;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * This class contains common behavior for a RSK native contract.
+ *
+ * @author Ariel Mendelzon
+ */
+public abstract class NativeContract extends PrecompiledContracts.PrecompiledContract {
+    private static final Logger logger = LoggerFactory.getLogger(NativeContract.class);
+    private static final PanicProcessor panicProcessor = new PanicProcessor();
+
+    private final RskSystemProperties config;
+
+    private ExecutionEnvironment executionEnvironment;
+
+    public NativeContract(RskSystemProperties config, RskAddress contractAddress) {
+        this.contractAddress = contractAddress;
+        this.config = config;
+    }
+
+    public ExecutionEnvironment getExecutionEnvironment() {
+        return executionEnvironment;
+    }
+
+    public abstract List<NativeMethod> getMethods();
```

Comment:
I think it would be better if the methods were passed via constructor.

Reply:
Why would you pass them via constructor?

Reply:
That is a good point, actually. This method is not actually used publicly, so it can indeed be removed.

The problem with building methods at construction time is that they receive an execution environment. Removing that bit would mean that one would have to either:

1 - Manually set the execution environment before operating on a method instance, which is a pattern I don't personally like very much. (`method.setEnvironment(); method.doSomething();`).
2 - Clutting every method function with an additional parameter, the execution environment itself.

When designing this, the fact that every operation actually needed to be aware of its environment to do anything seemed like a good enough reason to set it as part of its state.